src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php

```diff
@@ -0,0 +1,225 @@
+                    }
+                    $this->addProductValueToCollections($value, $productValues);
+                }
+            }
+        }
+
+        return $productValues;
+    }
+
+    /**
+     * Add the (empty) required product value to the right product value collections depending on the localizable and
+     * scopable characteristics of product value.
+     *
+     * @param ProductValueInterface               $value
+     * @param ProductValueCollectionInterface[][] $productValueCollections
+     */
+    protected function addProductValueToCollections(ProductValueInterface $value, array $productValueCollections)
+    {
+        foreach ($productValueCollections as $channelCode => $productValueCollectionsByChannel) {
+            foreach ($productValueCollectionsByChannel as $localeCode => $productValueCollection) {
+                if ((null === $value->getScope() || null !== $channelCode = $value->getScope()) &&
+                    (null === $value->getLocale() || null !== $localeCode = $value->getLocale())
+                ) {
+                    $productValueCollection->add($value);
+                }
+            }
+        }
+    }
+
+    /**
+     * Generates one completeness for given required product value, channel code,
+     * locale code and the product values to compare.
+     *
+     * @param ProductInterface                $product
+     * @param ProductValueCollectionInterface $requiredValues
+     * @param ProductValueCollectionInterface $actualValues
+     * @param string                          $channelCode
+     * @param string                          $localeCode
+     *
+     * @return CompletenessInterface
+     */
+    protected function generateCompleteness(
+        ProductInterface $product,
+        ProductValueCollectionInterface $requiredValues,
+        ProductValueCollectionInterface $actualValues,
+        $channelCode,
+        $localeCode
+    ) {
+        $channel = $this->channelRepository->findOneByIdentifier($channelCode);
+        $locale = $this->localeRepository->findOneByIdentifier($localeCode);
+
+        $missingAttributes = new ArrayCollection();
+        $missingCount = 0;
+        $requiredCount = 0;
+
+        foreach ($requiredValues as $requiredValue) {
+            $productValue = $actualValues->getByCodes(
+                $requiredValue->getAttribute()->getCode(),
+                $requiredValue->getScope(),
+                $requiredValue->getLocale()
+            );
+
+            if (null === $productValue ||
+                !$this->productValueCompleteChecker->isComplete($productValue, $channel, $locale)
+            ) {
+                $attribute = $requiredValue->getAttribute();
+
+                if (!$missingAttributes->contains($attribute)) {
+                    $missingAttributes->add($attribute);
+                    $missingCount++;
+                }
+            }
+
+            $requiredCount++;
+        }
+
+        $completeness = new Completeness(
```

Comment:
We could. To see if it is usefull as we don't create completenesses anywhere else than here.